# Mini documentation for each tools
### Specify the inputs and outputs for each cwl script
- ## salmon_count
   #### inputs:  
   `input_script`: R count script  
   `gtf`: annotation file in gtf format  
   `metadata`: metadata.csv  
   `salmon_dir`: directory with subdirectorys of outputs from `salmon_quant`  
   > maybe change this directory name? like `quant_results`?
   #### outputs:
   `gene_abundance_matrix.csv`  
   `gene_count_matrix.csv` * this one used for DGE analysis e.g. DESeq2  
   `gene_length_matrix.csv`

- ## ballgown
   #### inputs:
   `script`: R ballgown script.
   > should be changed to input_script
   `tablemaker_output`: directory of the tablemaker output
   `metadata`: metadata csv file.
   `condition`: string giving the condition of interest in the metadata columns
   #### outputs:
   `DGE_res.csv`
   `DTE_res.csv`

- ## cuffdiff
  #### inputs:
  `output`: name for output directory
  `threads`: number of cores to use
  `label`: the factors of the condition of interest
  `FDR`: the FDR rate to which to label as significant, could default to 0.05.
  `gtf_file`: merged gtf annotation file in gtf format
  > should label as 'merged_gtf' to keep consistent
  `condition1_files`: cxb files of all samples for condition 1. cuffquant output
  `condition2_files`: cxb files of all samples for condition 2. cuffquant output
  #### outputs:
  `output`: names after inputs.output. directory containing all cuffdiff output files
  > can add gene_exp.diff output only.

- ## cufflinks
  #### inputs:
  `gtf`: annotation file in gtf format
  `output_dir`: name for output directory
  > should change to just `output`
  `threads`: number of threads to use.
  `alignment_file`: bam file
  > should change to `bam`
  #### outputs:
  `output`: named after output_dir. directory with all cufflink file outputs
  `transcripts.gtf`

- ## cuffmerge:
  #### inputs:
  `output`: string stating the output directory name
  `gtf`: annotation file in gtf format
  `threads`: number of threads to use.
  `ref_fasta`: fasta file used in indexing
  > should change to `fasta`
  `cufflinks_output`: transcripts.gtf files generated by cufflinks
  #### outputs:
  `output`: output directory named after input.output
  `merged.gtf`

- ## cuffnorm:
  #### inputs:
  `output`: name for output directory
  `threads`: number of threads to use.
  `gtf_file`: merged gtf annotation file in gtf format
  > should label as 'merged_gtf' to keep consistent
  `condition1_files`: cxb files of all samples for condition 1. cuffquant output
  `condition2_files`: cxb files of all samples for condition 2. cuffquant output
  #### outputs:
  `output`: named after inputs.output. file contained normalised gene count matrix

- ## cuffquant:
  #### inputs:
  `output`: name for output directory
  `threads`: number of threads to use.
  `merged_gtf`: merged gtf annotation file in gtf format
  `alignment_file`: bam file
  > should change to `bam`
  #### outputs:
  `output`: named after inputs.output. Directory containing all quant files
  `abundances.cxb`

- ## DESeq2:
  #### inputs:
  `script`: R DESeq2 script.
  > should be changed to input_script
  `count_matrix`: gene count matrix file
  > change to `count_matrix`
  `metadata`: metadata csv file.
  #### outputs:
  `DGE_results.csv`
  > change to DGE_res.csv

- ## DEXSeq:
  #### inputs:
  `input_script`: R DEXSeq script
  `count_matrix`: exon count matrix file
  `gff`: directory containing gff file from htseq prepare
  `metadata`: metadata csv file.
  #### outputs:
  `DEE_results.csv`
  > change to `DEE_res.csv`

- ## edger
  #### inputs:
  `script`: R edger script
  > change to `input_script`
  `condition`: string giving the condition of interest in the metadata columns
  `counts`: gene counts matrix
  > change to `count_matrix`
  `metadata`: metadata csv file.
  #### outputs:
  `DGE_res.csv`
