# Mini documentation for each tools
### Specify the inputs and outputs for each cwl script

- ## ballgown
   #### inputs:
   `input_script`: R ballgown script.  
    > was script  

   `tablemaker_output`: directory of the tablemaker output  
   `metadata`: metadata csv file.  
   `condition`: string giving the condition of interest in the metadata columns  
   #### outputs:
   `DGE_res.csv`  
   `DTE_res.csv`

- ## cuffdiff
  #### inputs:
  `output`: name for output directory  
  `threads`: number of cores to use  
  `label`: the factors of the condition of interest  
  `FDR`: the FDR rate to which to label as significant, could default to 0.05.  
  `merged_gtf`: merged gtf annotation file in gtf format  
  > was 'gtf_file' 

  `condition1_files`: cxb files of all samples for condition 1. cuffquant output  
  `condition2_files`: cxb files of all samples for condition 2. cuffquant output  
  #### outputs:
  `output`: names after inputs.output. directory containing all cuffdiff output files  
  > can add gene_exp.diff output only.

- ## cufflinks
  #### inputs:
  `gtf`: annotation file in gtf format  
  `output`: name for output directory  
  > was `output_dir`

  `threads`: number of threads to use.  
  `bam`: bam file  
  > was `alignment_file`

  #### outputs:
  `output`: named after output_dir. directory with all cufflink file outputs  
  `transcripts.gtf`

- ## cuffmerge:
  #### inputs:
  `output`: string stating the output directory name  
  `gtf`: annotation file in gtf format  
  `threads`: number of threads to use.  
  `fasta`: fasta file used in indexing  
  > was `ref_fasta` 

  `cufflinks_output`: transcripts.gtf files generated by cufflinks  
  #### outputs:
  `output`: output directory named after input.output  
  `merged.gtf`  

- ## cuffnorm:
  #### inputs:
  `output`: name for output directory  
  `threads`: number of threads to use.  
  `merged_gtf`: merged gtf annotation file in gtf format  
  > was 'gtf_file'

  `condition1_files`: cxb files of all samples for condition 1. cuffquant output  
  `condition2_files`: cxb files of all samples for condition 2. cuffquant output  
  #### outputs:
  `output`: named after inputs.output. file contained normalised gene count matrix

- ## cuffquant:
  #### inputs:
  `output`: name for output directory  
  `threads`: number of threads to use.  
  `merged_gtf`: merged gtf annotation file in gtf format  
  `bam`: bam file  
  > was `alignment_file`

  #### outputs:
  `output`: named after inputs.output. Directory containing all quant files  
  `abundances.cxb`

- ## DESeq2:
  #### inputs:
  `input_script`: R DESeq2 script.  
  > `script`

  `count_matrix`: gene count matrix file  
  `metadata`: metadata csv file.  
  #### outputs:
  `DGE_results.csv`  

- ## DEXSeq:
  #### inputs:
  `input_script`: R DEXSeq script  
  `count_matrix`: exon count matrix file  
  `gff`: directory containing gff file from htseq prepare  
  `metadata`: metadata csv file.  
  `threads`: number of threads to use
  #### outputs:
  `DEE_results.csv`  
  > change to `DEE_res.csv`

- ## edger:
  #### inputs:
  `input_script`: R edger script  
  > was `script`

  `condition`: string giving the condition of interest in the metadata columns  
  `count_matrix`: gene counts matrix  
  > was `counts`

  `metadata`: metadata csv file.  
  #### outputs:
  `DGE_res.csv`

- ## featurecounts:
  #### inputs:
  `input_script`: R featurecounts script  
  > was `script`

  `bam_files`: all bam files  
  `gtf`: annotation in ftd format  
  `threads`: number of threads to use.  
  `metadata`: used to see the libType of each sample.  
  #### outputs:
  `gene_count_matrix.csv`

- ## fgsea:
  #### inputs:
  `input_script`: R fgsea input_script  
  > change to `input_scripts` <

  `de_results`: DGE res file  
  > was `de_res`

  `gene_set`: file containing gene set information in long form.  
  #### ouputs:
  `gsea_res.csv`

- ## hisat2_align:
  #### inputs:
  `threads`: number of threads  
  `index_directory`: path to hisat2 index  
  `first_pair`: first fastq file  
  `second_pair`: second fastq file  
  `output`: output name  
  > was `sam_name`

  `XSTag`: Tag to use ("--dta" or "--dta-cufflinks")  

- ## hisat2_build:
  #### inputs:
  `fasta`: fasta file  
  > was `reference`

  `threads`: number of threads to use.  
  `output`: name to use for output  
  > was `basename` <

  #### outputs:
  `output`: samfile with a basename given from inputs.output  
  `log`

- ## htseq_count:
  #### inputs:
  `input_script`: python htseq_count script  
  `pairedend`: logical  
  `stranded`: logical  
  `input_format`: bam or sam  
  `sorted_by`: pos  
  `gff`: gff file from htseq_annotation  
  `bam`: bam or sam file  
  > was `sam`

  `outname`: name for outfile
  #### outputs:
  `outname`: named after input.ouput  

- ## htseq_prepare:
  #### inputs:
  `input_script`: python input script  
  `gtf`: gtf annotation file  
  `gff_name`: name for gff output  
  #### outputs:
  `gff_name`: gff file named after inputs.gff_name

- ## hypergeo
  #### inputs:
  `input_script`: R HyperGeo_Script script  
  > was `script`

  `de_res`: DGE res file  
  `gene_set`: file containing gene set information in long form.  
  #### outputs:
  `hypergeo_res.csv`

- ## miso_index:
  #### inputs:
  `gtf`: gtf file  
  > was `annotation`

  `output`: output name for output directory  
  > was `index_dir` <
  #### outputs:
  `index_dir`: output directory named after inputs.index_dir

- ## miso_run
  #### inputs:
  `threads`: number of threads to use  
  > was 'cores' <

  `lib_type`: pairedend or not  
  > change to `pairedend` or change all other to `lib_type`

  `index_directory`: Directory from miso_index output  
  > was `index`

  `bam`: bam file  
  `read_len`: read length  
  `gtf`: gtf file  
  > was `annotation_file`

  `min_exon_size`: minimum exon size to use  
  `output`: output name  
  > was out_dir
  #### outputs:
  `out_dir`: directory named after inputs.out_dir

- ## prepDE
  #### inputs:
  `input_script`: python prepDE script  
  > was `program

  `stringtie_out`: gtf files from stringtie  
  > was `gtfs`
  #### outputs:
  `gene_count_matrix.csv`
  `transcript_count_matrix.csv`

- ## salmon_count
  #### inputs:  
  `input_script`: R count script  
  `gtf`: annotation file in gtf format  
  `metadata`: metadata.csv  
  `quant_results`: directory with subdirectorys of outputs from `salmon_quant`  
  > was `salmon_dir`
  #### outputs:
  `gene_abundance_matrix.csv`  
  `gene_count_matrix.csv` * this one used for DGE analysis e.g. DESeq2  
  `gene_length_matrix.csv`

- ## salmon_index:
  #### inputs:
  `fasta`: fasta files  
  `index_type`: 2 different type of running "fmd" or "quasi"  
  `threads`: number of threads to use  
  > was `cores`

  `output`: directory output name  
  > was `index_name`
  #### outputs:
  `index_name`: output directory with name inputs.index_name

- ## salmon_quant:
  #### inputs:
  `index_directory`: directory of salmon index  
  > was `index`

  `threads`: number of threads to use  
  > was `cores`

  `output`: output name  
  > was `out_dir`

  `first_end_fastq`: if paired end. first pair fastq file  
  `second_end_fastq` if paired end. second pair fastq file  
  `single_fastq`: if single end. single fastq file  
  #### outputs:
  `out_dir`: Directory with name of inputs.out_dir

- ## samtools:
  #### inputs:
  `action`: samtools command (default in sort)  
  `sortby`: how to sort the bam file  
  `threads`: how many threads to use (total - 1), it is additional  
  `samfile`: sam file  
  > change to `sam`

  `outfilename`: ouutput file name  
  > change to `output`
  #### outputs:
  `outfilename`: bam file with name inputs.outfilename

- ## STAR_index:
  #### inputs:
  `threads`: number of threads  
  `Mode`: how to run STAR (use genomeGenerate)  
  `output`: directory name to use  
  > was `genomeDir`

  `fasta`: fasta files  
  > was `genomeFastaFiles`

  `gtf`: gtf file
  > was `sjdbGTFfile`

- ## STAR_readmap
  #### inputs:
  `threads`: number of threads to use  
  `Mode`: how to run STAR (use alignReads)  
  `genomeDir`: directory name to use  
  `sjdbGTFfile`: gtf file  
  > change to `gtf`

  `sjdbOverhang`  
  `readFilesIn`: fastq files  
  `outSAMtype`: state the output wanted use("BAM SortedByCoordinate")  
  `outFileNamePrefix`: output name for directory  
  > change to `output`
  #### outputs:
  `outFileNamePrefix`: output directory name with name inputs.outFileNamePrefix  
  `sam`: sam file  

- ## stringtie
  #### inputs:
  `bam`: bam file  
  > was `input_bam`

  `threads`: number of threads to use.  
  `gtf`: gtf file  
  > was `annotation`

  `output`: output name for file  
  > was `outfilename`
  #### outputs:
  `outfilename`: output file name with name inputs.outfilename

- ## tablemaker:
  #### inputs:
  `threads`: number of threads to use.  
  `merged_gtf`: merged gtf from cuffmerge  
  `bam`: bam files to use  
  `output`: output name for directory.  
  #### outputs:
  `output`: Directory with name inputs.outputs
